atomics/include/desul/atomics/Compare_Exchange_HIP.hpp

```diff
@@ -17,130 +18,72 @@ SPDX-License-Identifier: (BSD-3-Clause)
-                "this function assumes an unsigned long long is 64-bit");
-  unsigned long long int return_val =
-      atomicExch(reinterpret_cast<unsigned long long int*>(dest),
-                 reinterpret_cast<unsigned long long int&>(value));
-  return reinterpret_cast<T&>(return_val);
+  T return_val = __hip_atomic_exchange(dest, value,
+                                       HIPMemoryOrder<MemoryOrderRelaxed>::value,
+                                       HIPMemoryScope<MemoryScope>::value);
+  return return_val;
 }
 
 template <class T, class MemoryScope>
-__device__ std::enable_if_t<sizeof(T) == 4 || sizeof(T) == 8, T> device_atomic_exchange(
+__device__ std::enable_if_t<atomic_exchange_available_hip<T>::value, T> device_atomic_exchange(
     T* const dest, T compare, T value, MemoryOrderRelease, MemoryScope) {
   T return_val = device_atomic_compare_exchange(
-      dest, compare, value, MemoryOrderRelaxed(), MemoryScope());
-  device_atomic_thread_fence(MemoryOrderRelease(), MemoryScope());
+      dest, compare, value, MemoryOrderRelease(), MemoryScope());
   return reinterpret_cast<T&>(return_val);
 }
 
 template <class T, class MemoryScope>
-__device__ std::enable_if_t<sizeof(T) == 4 || sizeof(T) == 8, T> device_atomic_exchange(
+__device__ std::enable_if_t<atomic_exchange_available_hip<T>::value, T> device_atomic_exchange(
     T* const dest, T /*compare*/, T value, MemoryOrderAcquire, MemoryScope) {
-  device_atomic_thread_fence(MemoryOrderAcquire(), MemoryScope());
-  T return_val =
+  T return_val = __hip_atomic_exchange(dest, value,
+                                       HIPMemoryOrder<MemoryOrderAcquire>::value,
+                                       HIPMemoryScope<MemoryScope>::value);
       device_atomic_exchange(dest, value, MemoryOrderRelaxed(), MemoryScope());
-  return reinterpret_cast<T&>(return_val);
+  return return_val;
 }
 
 template <class T, class MemoryScope>
-__device__ std::enable_if_t<sizeof(T) == 4 || sizeof(T) == 8, T> device_atomic_exchange(
+__device__ std::enable_if_t<atomic_exchange_available_hip<T>::value, T> device_atomic_exchange(
     T* const dest, T value, MemoryOrderAcqRel, MemoryScope) {
-  device_atomic_thread_fence(MemoryOrderAcquire(), MemoryScope());
-  T return_val =
-      device_atomic_exchange(dest, value, MemoryOrderRelaxed(), MemoryScope());
-  device_atomic_thread_fence(MemoryOrderRelease(), MemoryScope());
-  return reinterpret_cast<T&>(return_val);
+  T return_val = __hip_atomic_exchange(dest, value,
+                                       HIPMemoryOrder<MemoryOrderAcqRel>::value,
+                                       HIPMemoryScope<MemoryScope>::value);
+  return return_val;
 }
 
 template <class T, class MemoryScope>
-__device__ std::enable_if_t<sizeof(T) == 4 || sizeof(T) == 8, T> device_atomic_exchange(
+__device__ std::enable_if_t<atomic_exchange_available_hip<T>::value, T> device_atomic_exchange(
     T* const dest, T value, MemoryOrderSeqCst, MemoryScope) {
```

Comment:
Why would we provide this more specialized version?
Is it not covered by the generic overload L41?

Reply:
Did you mean to have all these `device_atomic_exchange` overload L49-L83 
https://github.com/desul/desul/blob/84cfa75787c2e2deafaa801fa3a45f8ad73c2c7a/atomics/include/desul/atomics/Compare_Exchange_HIP.hpp#L49-L83
or did you forget to remove?